File: 0501-1000/0509_fib.py
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1: return n
        prev = 0
        curr = 1
        for _ in range(n-1):
            curr, prev = curr + prev, curr
        return curr

# Alternative solution

class Solution:
    def fib(self, n: int) -> int:
        self.memo = {}
        return self.dp(n)

    def dp(self, n):
        if n <= 1:
            return n
        
        if n in self.memo:
            return self.memo[n]

        self.memo[n] = self.dp(n-1) + self.dp(n-2)
        return self.memo[n]

File: 0001-0500/0509_fib.java
class Solution {
    public int fib(int n) {
        if (n < 2) return n;
        
        int first = 0;
        int second = 1;
        int curr;

        for (int i=2; i<n+1; i++) {
            curr = first + second;
            first = second;
            second = curr;
        }

        return second;
    }
}
